src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -173,14 +231,19 @@ public Task<bool> WriteStreamAsync(TKey key, Stream stream, Checksum? checksum,
 
             private bool WriteStream(TKey key, Stream stream, Checksum? checksum, CancellationToken cancellationToken)
             {
+                // We're writing.  This better always be under the exclusive scheduler.
+                Contract.ThrowIfFalse(TaskScheduler.Current == Storage._connectionPoolService.Scheduler.ExclusiveScheduler);
+
                 cancellationToken.ThrowIfCancellationRequested();
 
                 if (!Storage._shutdownTokenSource.IsCancellationRequested)
                 {
                     using var _ = Storage._connectionPool.Target.GetPooledConnection(out var connection);
 
-                    // Determine the appropriate data-id to store this stream at.
-                    if (TryGetDatabaseId(connection, key, out var dataId))
+                    // Determine the appropriate data-id to store this stream at.  We already are running
+                    // with an exclusive write lock on the DB, so it's safe for us to write the data id to 
+                    // the db on this connection if we need to.
+                    if (TryGetDatabaseId(connection, key, allowWrite: true, out var dataId))
```

Comment:
this is still safe to have here in a single call (as explained in the comment above).